[dynamo][numpy] Add support for np.dtype

larryliu0820 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103546

## Problem

Trying to support numpy function call in dynamo, with numpy dtype as argument.

For example:

```
def fn(x: int):
    return np.empty_like(x, dtype=np.float64)
```

## Solution

This currently doesn't work because `NumpyVariable` doesn't implement `as_proxy()`. The idea in `as_proxy()` for now is to convert `np.float64` and other np.<dtype> into `torch.dtype` and then feed into the corresponding `torch_np` method.

For previous example, we convert `np.float64` to `torch.float64` in `as_proxy()` and then feed it into `torch_np.empy_like()` method.
cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @msaroufim

larryliu0820 committed:
[dynamo][numpy] Add support for np.dtype

[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo][numpy] Add support for np.dtype"




cc voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim

[ghstack-poisoned]

lezcano commented:
@ev-br which other `np.` objects may appear in a program? All these will need to be handled by Dynamo in a similar way.

ev-br commented:
```
>>> import torch_np as tnp
>>> [x for x in dir(tnp) if not x.startswith('_')]
['AxisError', 'DType', 'False_', 'True_', 'UFuncTypeError', 'abs', 'absolute', 'add', 'all', 'allclose',
 'alltrue', 'amax', 'amin', 'angle',  'any', 'append', 'arange', 'arccos', 'arccosh', 'arcsin', 'arcsinh',
 'arctan', 'arctan2', 'arctanh', 'argmax', 'argmin', 'argsort',  'argwhere', 'around', 'array', 
'array_equal', 'array_equiv', 'array_split', 'asarray', 'ascontiguousarray', 'asfarray',  'atleast_1d', 
'atleast_2d', 'atleast_3d', 'average', 'bartlett', 'bincount', 'bitwise_and', 'bitwise_not', 'bitwise_or', 
'bitwise_xor', 'blackman', 'block', 'bool_', 'broadcast_arrays', 'broadcast_shapes', 'broadcast_to', 
'byte', 'can_cast', 'cbrt', 'cdouble', 'ceil', 'choose', 'clip', 'column_stack', 'common_type',
'complex128', 'complex64', 'complex_', 'complexfloating', 'concatenate',  'conj', 'conjugate',
'convolve', 'copy', 'copysign', 'copyto', 'corrcoef', 'correlate', 'cos', 'cosh', 'count_nonzero', 'cov',
'cross',  'csingle', 'cumprod', 'cumproduct', 'cumsum', 'deg2rad', 'degrees', 'diag', 'diag_indices',
 'diag_indices_from', 'diagflat',  'diagonal', 'diff', 'divide', 'divmod', 'dot', 'double', 'dsplit', 'dstack',
'dtype', 'e', 'einsum', 'empty', 'empty_like', 'equal', 'errstate', 'exp', 'exp2', 'expand_dims', 'expm1',
'eye', 'fabs', 'fft', 'fill_diagonal', 'finfo', 'fix', 'flatnonzero', 'flip', 'fliplr',  'flipud', 'float16', 'float32',
 'float64', 'float_', 'float_power', 'floating', 'floor', 'floor_divide', 'fmax', 'fmin', 'fmod', 'from_dlpack',
'full', 'full_like', 'gcd', 'generic', 'geomspace', 'gradient', 'greater', 'greater_equal', 'half', 'hamming',
'hanning', 'heaviside', 'histogram', 'histogram2d', 'histogramdd', 'hsplit', 'hstack', 'hypot', 'i0',
'identity', 'iinfo', 'imag', 'index_exp', 'indices', 'inexact', 'inf', 'inner', 'int16', 'int32', 'int64', 'int8',
'int_', 'intc', 'integer', 'intp', 'invert', 'isclose', 'iscomplex', 'iscomplexobj', 'isfinite', 'isinf', 'isnan',
'isneginf', 'isposinf', 'isreal', 'isrealobj', 'isscalar', 'issubdtype', 'kaiser', 'kron', 'lcm', 'ldexp',
'left_shift', 'less', 'less_equal', 'linalg', 'linspace', 'log', 'log10', 'log1p', 'log2', 'logaddexp',
'logaddexp2', 'logical_and', 'logical_not', 'logical_or', 'logical_xor', 'logspace', 'longlong',
'matmul', 'max', 'maximum', 'mean', 'median', 'meshgrid', 'min', 'min_scalar_type', 'minimum',
'mintypecode', 'mod', 'modf', 'moveaxis', 'multiply', 'nan', 'nan_to_num', 'nanargmax', 'nanargmin',
'nancumprod', 'nancumsum', 'nanmax', 'nanmean', 'nanmedian', 'nanmin', 'nanpercentile',
'nanprod', 'nanquantile', 'nanstd', 'nansum', 'nanvar', 'ndarray', 'ndim', 'negative', 'newaxis',
'nextafter', 'nonzero', 'not_equal', 'number', 'ones', 'ones_like', 'outer', 'pad', 'percentile', 'pi',
'positive', 'power', 'prod', 'product', 'ptp', 'put', 'put_along_axis', 'quantile', 'rad2deg', 'radians',
'random', 'ravel', 'real', 'real_if_close', 'reciprocal', 'remainder', 'repeat', 'reshape', 'resize',
'result_type', 'right_shift', 'rint', 'roll', 'rollaxis', 'rot90', 'round', 'round_', 'row_stack', 's_',
'sctypes', 'searchsorted', 'set_default_dtype', 'shape', 'short', 'sign', 'signbit', 'signedinteger',
'sin', 'sinc', 'single', 'sinh', 'size', 'sometrue', 'sort', 'split', 'sqrt', 'square', 'squeeze', 'stack', 'std',
'subtract', 'sum', 'swapaxes', 'take', 'take_along_axis', 'tan', 'tanh', 'tensordot', 'tile', 'trace',
'transpose', 'tri', 'tril', 'tril_indices', 'tril_indices_from', 'triu', 'triu_indices', 'triu_indices_from',
'true_divide', 'trunc', 'typecodes', 'ubyte', 'uint8', 'unique', 'unsignedinteger', 'vander', 'var',
'vdot', 'vsplit', 'vstack', 'where', 'zeros', 'zeros_like']
```

All these names by design match numpy, save:

```
>>> import numpy as np
>>> [x for x in set(dir(tnp)) - set(dir(np)) if not x.startswith('_')]
['UFuncTypeError', 'set_default_dtype', 'DType']
```

ev-br commented:
Python types, certainly, as in `np.eye(3, dtype=float)`. Else, in our subset of numpy, there mostly dtype aliases (abc.ABCMeta below), plus a couple of exceptions (those can be cleaned out) and a couple of classes : 

```
In [26]: def pred(name):
    ...:     obj = getattr(tnp, name)
    ...:     return inspect.isfunction(obj) or inspect.ismodule(obj) or name.startswith('_')
    ...: 

In [27]: [(x, type(getattr(tnp, x))) for x in dir(tnp) if not pred(x)]
Out[27]: 
[('AxisError', type),
 ('DType', type),
 ('False_', torch_np._ndarray.ndarray),
 ('True_', torch_np._ndarray.ndarray),
 ('UFuncTypeError', type),
 ('bool_', abc.ABCMeta),
 ('byte', abc.ABCMeta),
 ('cdouble', abc.ABCMeta),
 ('complex128', abc.ABCMeta),
 ('complex64', abc.ABCMeta),
 ('complex_', abc.ABCMeta),
 ('complexfloating', abc.ABCMeta),
 ('csingle', abc.ABCMeta),
 ('double', abc.ABCMeta),
 ('e', float),
 ('float16', abc.ABCMeta),
 ('float32', abc.ABCMeta),
 ('float64', abc.ABCMeta),
 ('float_', abc.ABCMeta),
 ('floating', abc.ABCMeta),
 ('generic', abc.ABCMeta),
 ('half', abc.ABCMeta),
 ('index_exp', torch_np._funcs.IndexExpression),
 ('inexact', abc.ABCMeta),
 ('inf', float),
 ('int16', abc.ABCMeta),
 ('int32', abc.ABCMeta),
 ('int64', abc.ABCMeta),
 ('int8', abc.ABCMeta),
 ('int_', abc.ABCMeta),
 ('intc', abc.ABCMeta),
 ('integer', abc.ABCMeta),
 ('intp', abc.ABCMeta),
 ('longlong', abc.ABCMeta),
 ('nan', float),
 ('ndarray', type),
 ('newaxis', NoneType),
 ('number', abc.ABCMeta),
 ('pi', float),
 ('s_', torch_np._funcs.IndexExpression),
 ('sctypes', dict),
 ('short', abc.ABCMeta),
 ('signedinteger', abc.ABCMeta),
 ('single', abc.ABCMeta),
 ('typecodes', dict),
 ('ubyte', abc.ABCMeta),
 ('uint8', abc.ABCMeta),
 ('unsignedinteger', abc.ABCMeta)]
```

ev-br commented:
Also, I don't know if it's relevant or not:  our `tnp.dtype` is a regular function. In numpy proper, it's something else, `type(np.dtype) -> numpy._DTypeMeta`.

larryliu0820 committed:
Update on "[dynamo][numpy] Add support for np.dtype"


## Problem

Trying to support numpy function call in dynamo, with numpy dtype as argument.

For example:

```
def fn(x: int):
    return np.empty_like(x, dtype=np.float64)
```

## Solution

This currently doesn't work because `NumpyVariable` doesn't implement `as_proxy()`. The idea in `as_proxy()` for now is to convert `np.float64` and other np.<dtype> into `torch.dtype` and then feed into the corresponding `torch_np` method.

For previous example, we convert `np.float64` to `torch.float64` in `as_proxy()` and then feed it into `torch_np.empy_like()` method.
cc voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim

[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo][numpy] Add support for np.dtype"


## Problem

Trying to support numpy function call in dynamo, with numpy dtype as argument.

For example:

```
def fn(x: int):
    return np.empty_like(x, dtype=np.float64)
```

## Solution

This currently doesn't work because `NumpyVariable` doesn't implement `as_proxy()`. The idea in `as_proxy()` for now is to convert `np.float64` and other np.<dtype> into `torch.dtype` and then feed into the corresponding `torch_np` method.

For previous example, we convert `np.float64` to `torch.float64` in `as_proxy()` and then feed it into `torch_np.empy_like()` method.
cc voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim

[ghstack-poisoned]

ezyang reviewed (changes requested):
Converting the Numpy dtype into a Torch dtype seems like the wrong way to go about doing this, because I expect that torch_np compatibility wrapper to handle this. Can you help me understand what goes wrong if we don't do a conversion? Does FX complain that np.dtype is not a valid argument in FX?

larryliu0820 commented:
> Converting the Numpy dtype into a Torch dtype seems like the wrong way to go about doing this, because I expect that torch_np compatibility wrapper to handle this. Can you help me understand what goes wrong if we don't do a conversion? Does FX complain that np.dtype is not a valid argument in FX?

Yeah, np.float64 will be wrapped into a NumpyVariable and there's no way to turn it into a proxy. Even after I hack around that logic, torch_np doesn't take np.dtype as an argument.

ezyang commented:
> Even after I hack around that logic, torch_np doesn't take np.dtype as an argument.

Uhh, how come? @ev-br, @Lezcano, do you know why?

larryliu0820 commented:
> > Even after I hack around that logic, torch_np doesn't take np.dtype as an argument.
> 
> Uhh, how come? @ev-br, @lezcano, do you know why?

I think it’s by design, torch_np doesn’t take any bumpy stuff.

lezcano commented:
We implement all the NumPy namespace so, in the same way that `torch_np.array` returns a `torch_np._ndarray`, `torch_np.dtype` returns our own implementation of `np.dtype`. If you see an `np.dtype` and rather than using that, you call `torch_np.dtype` and you feed that to the function, everything should work.

ezyang commented:
Let's do what @lezcano says.

larryliu0820 commented:
> Let's do what @lezcano says.

Though torch_np.dtype can't be converted into a proxy, it will run into https://github.com/pytorch/pytorch/blob/main/torch/fx/proxy.py#L271.

I can add torch_np dtype into a proxy-recognized `base_types` to workaround it.

larryliu0820 commented:
> Let's do what @lezcano says.

Actually converting numpy.dtype to string works here.

larryliu0820 committed:
Update on "[dynamo][numpy] Add support for np.dtype"


## Problem

Trying to support numpy function call in dynamo, with numpy dtype as argument.

For example:

```
def fn(x: int):
    return np.empty_like(x, dtype=np.float64)
```

## Solution

This currently doesn't work because `NumpyVariable` doesn't implement `as_proxy()`. The idea in `as_proxy()` for now is to convert `np.float64` and other np.<dtype> into `torch.dtype` and then feed into the corresponding `torch_np` method.

For previous example, we convert `np.float64` to `torch.float64` in `as_proxy()` and then feed it into `torch_np.empy_like()` method.
cc voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim

[ghstack-poisoned]

lezcano reviewed (commented):
How does this work?
```python
>>> import numpy as np
>>> np.dtype(int).__name__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'numpy.dtype[int64]' object has no attribute '__name__'. Did you mean: '__ne__'?
```

Should this be `__str__()`? And in that case, it'd be better to simply call `str()`.

larryliu0820 commented:
> How does this work?
> 
> ```python
> >>> import numpy as np
> >>> np.dtype(int).__name__
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> AttributeError: 'numpy.dtype[int64]' object has no attribute '__name__'. Did you mean: '__ne__'?
> ```
> 
> Should this be `__str__()`? And in that case, it'd be better to simply call `str()`.

```python
>>> import numpy as np
>>> np.float32.__name__
'float32'
>>> type(np.float32)
<class 'type'>
```

str() doesn't work though:

```python
>>> str(np.float32)
"<class 'numpy.float32'>"
```

lezcano commented:
In that case,

```python
>>> np.empty(3, dtype=np.dtype(float))
array([4.63460416e-310, 0.00000000e+000, 2.44011711e-154])
```
this works and would not work with your solution?

lezcano commented:
Is it not possible to fetch all these objects from the `torch_np` to make sure we get the correct semantics in all cases?

larryliu0820 commented:
> Is it not possible to fetch all these objects from the `torch_np` to make sure we get the correct semantics in all cases?

So for graph module to take `torch_np.float64` args it's a lot of pain.. I'm not sure if we want to go that route. For one thing, `torch_np.float64` are all `abc.ABCMeta` type, this is not one of the [`BaseArgumentTypes`](https://github.com/pytorch/pytorch/blob/main/torch/fx/node.py#L17) for `node`.

Even after I add it, downstream I need to handle `abc.ABCMeta` to be correctly represented in the graph code, right now it just shows up as:

```
def forward(self, L_x_ : torch.Tensor):
    l_x_ = L_x_
    wrapped_empty_like = torch__dynamo_utils_wrapped_empty_like(l_x_, dtype = <class 'torch_np._dtypes.float64'>);  l_x_ = None
    return (wrapped_empty_like,)
```

@ezyang  can you give some guidance on whether this is the right approach?

ev-br commented:
We can make these abcmetas to regular classes if it helps

lezcano commented:
@ev-br could you put up a PR changing all these into regular classes? Would that unblock you @larryliu0820?

larryliu0820 commented:
Yeah - we can try changing abcmeta to `type` and add `type` as one of `BaseArgumentTypes`. I'm not sure whether this will break something else so would love to hear @ezyang 's thoughts

lezcano commented:
Ed is on PTO for a couple weeks, so it'd be good if we can have something that unblocks, even if we have to revisit it later following Ed's comments' once he's back.

larryliu0820 commented:
> 

Sounds good to me

ev-br commented:
> https://github.com/pytorch/pytorch/pull/103546#issuecomment-1620478697

https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/160

lezcano commented:
@larryliu0820 it'll be merged in about 20 mins, when the CI finishes running. Please ping us if there's anything else that may be blocking you.

ezyang commented:
Stop the presses!

This has nothing to do with ABCMeta.  And I will reject a PR that adds type to BaseArgumentTypes.

The root cause of the issue here is that FX graphs have restrictions on argument types to ensure that, for example, we are able to actually serialize FX graphs into Python code that can be run by the Python interpreter. In particular, anything that can show up as an argument in an FX graph must be in the BaseArgumentTypes whitelist, which means that we actually know how to serialize the argument type into source code. This requirement is not so important for PT2, as our FX graphs typically do not get serialized (although they do get serialized in the minifier). But they are pretty important for other users of FX.

When @wanchaol recently ran into a similar problem, I suggested to him a quick workaround, which does NOT require modifying FX code, is to create an on-the-fly partially applied function that already has the naughty arguments applied; so instead of having foo(tensor, torch_np.float32), you instead have foo_float32(tensor), which will pass the FX validation check. This is probably the most non-intrusive way to get things to work.

There are some more intrusive ways to get things to work. We could add torch_np.dtype to the set of allowed base types (this is what @anijain2305 did for OpOverload types). This is fairly awkward, however, because torch_np is a separate package and we cannot guarantee the identifier is even available. We could also *entirely suspend* the allowed base types test. I am loathe to do this globally for the other users, but perhaps PT2 might be OK with it, and you could specify this on a per graph basis? In a pinch, I probably would prefer the former over the latter.

cc @zdevito as the few original FX authors still around

cc @wconstab @wanchaol  @bdhirsh  as distributed tracing through FX was also running into similar problems.

cc @anijain2305 @jansel @zou3519 who most recently modified BaseArgumentTypes

lezcano commented:
Uh, this looks like a deeper issue than I expected. I have a counter-offer that would sidestep all these issues:
What about normalising `dtype` into a string, similar to what @larryliu0820 initially suggested, but doing it properly. We already have that machinery in `torch_np`. You can simply do `tnp.dtype(dtype).name` and that will return something like `int64` and so on. These strings are already a valid input for any dtype kwarg in NumPy (and `torch_np`, of course), so this would make for a valid serialisation.

lezcano commented:
Note. You can also retrieve the equivalent torch dtype as `tnp.dtype(dtype).torch_dtype`. Now, technically speaking, I think that strings are better, as NumPy functions also accept these, while they don't accept torch dtypes.

ezyang commented:
strings seem like a fine workaround

lezcano commented:
@larryliu0820 does this unblock this PR?

larryliu0820 commented:
> @larryliu0820 does this unblock this PR?

Yes I was on leave last week will update the PR soon.

larryliu0820 committed:
Update on "[dynamo][numpy] Add support for np.dtype"


## Problem

Trying to support numpy function call in dynamo, with numpy dtype as argument.

For example:

```
def fn(x: int):
    return np.empty_like(x, dtype=np.float64)
```

## Solution

This currently doesn't work because `NumpyVariable` doesn't implement `as_proxy()`. The idea in `as_proxy()` for now is to convert `np.float64` and other np.<dtype> into `torch.dtype` and then feed into the corresponding `torch_np` method.

For previous example, we convert `np.float64` to `torch.float64` in `as_proxy()` and then feed it into `torch_np.empy_like()` method.
cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim

[ghstack-poisoned]

lezcano commented on torch/_dynamo/variables/misc.py:
```diff
@@ -805,8 +805,10 @@ def call_function(
         options = VariableTracker.propagate([[self]], [args], [list(kwargs.values())])
-        # lookup method name in torch_np
-        if hasattr(torch_np, self.value.__name__):
+        # lookup method name in torch_np. Things like np.dtype(float) are not supported yet.
```
I think it should be fine not to do it in this PR, but its something that would be good to have.

larryliu0820 committed:
Update on "[dynamo][numpy] Add support for np.dtype"


## Problem

Trying to support numpy function call in dynamo, with numpy dtype as argument.

For example:

```
def fn(x: int):
    return np.empty_like(x, dtype=np.float64)
```

## Solution

This currently doesn't work because `NumpyVariable` doesn't implement `as_proxy()`. The idea in `as_proxy()` for now is to convert `np.float64` and other np.<dtype> into `torch.dtype` and then feed into the corresponding `torch_np` method.

For previous example, we convert `np.float64` to `torch.float64` in `as_proxy()` and then feed it into `torch_np.empy_like()` method.
cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/misc.py:
```diff
@@ -837,6 +841,22 @@ def python_type(self):
+            import torch_np
+
+            try:
+                attr = next(a for a in dir(np) if getattr(np, a) is self.value)
```
...surely there's a better way to do this, then walking every public identifier in numpy (also, nit, let's just say numpy here, no need to abbreviate...)

ezyang commented on torch/_dynamo/variables/misc.py:
```diff
@@ -837,6 +841,22 @@ def python_type(self):
+                attr = next(a for a in dir(np) if getattr(np, a) is self.value)
+                # get torch_np equivalent
+                tnp_dtype = torch_np.dtype(attr)
+                return tnp_dtype.name
```
It's worth commenting why the string conversion happens here

larryliu0820 committed:
Update on "[dynamo][numpy] Add support for np.dtype"


## Problem

Trying to support numpy function call in dynamo, with numpy dtype as argument.

For example:

```
def fn(x: int):
    return np.empty_like(x, dtype=np.float64)
```

## Solution

This currently doesn't work because `NumpyVariable` doesn't implement `as_proxy()`. The idea in `as_proxy()` for now is to convert `np.float64` and other np.<dtype> into `torch.dtype` and then feed into the corresponding `torch_np` method.

For previous example, we convert `np.float64` to `torch.float64` in `as_proxy()` and then feed it into `torch_np.empy_like()` method.
cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 msaroufim

[ghstack-poisoned]

larryliu0820 commented:
@pytorchbot merge

larryliu0820 commented:
@pytorchbot merge -f "unrelated error"

voznesenskym commented:
@pytorchbot revert 

This fails `python benchmarks/dynamo/torchbench.py --only hf_BigBird --accuracy --training`

voznesenskym commented:
@pytorchbot revert -m "Failed on bench, unclear why bench test did not run on CI" -c weird

This fails python benchmarks/dynamo/torchbench.py --only hf_BigBird --accuracy --training

voznesenskym commented on torch/_dynamo/variables/misc.py:
```diff
@@ -837,6 +841,24 @@ def python_type(self):
+                return tnp_dtype.name
+            except StopIteration:
+                unimplemented(f"Can't find {self.value} in numpy module!")
+        return self.value
```
This is incorrectly gated. The VT pattern is almost always to invoke super() to "fall" up through the hierarchy if a given VT does not know how to handle a call. This PR subtly changes the `as_proxy` behavior from 

```
    def as_proxy(self):
        raise NotImplementedError(str(self))
```

Which is always valid to invoke in dynamo, as that causes a graph break or an inlining stop. 

Returning self.value here subtly changes the behavior, and causes code that would previously safely graph break out to now proceed. This is why this failed CI.

voznesenskym replied:
Confirmed locally, changing the return to `return super().as_proxy()` makes it pass. Please re-PR with this FX, and please ensure you have a test case to cover the failing case we found. If you are unsure how to do it, pull the pattern out of the failing model.
